### DISCUSSION TITLE
現在オープンのIssueを整理したい

### DISCUSSION THREAD

**Discussion by t1nyb0x** [misskey-dev/misskey#13964] [Ideas]
いつもお世話になっております。
普段IssueにProjectsの設定を行っているt1nyb0xです。

今回表題についてプロジェクトメンバー・コントリビューターの皆様からのご意見をいただきたくこのDiscussionを作りました。

## 提案内容

現在約2000件のIssueがオープンとなっているが、放置されているIssueは一度クローズする形で整理したい

## 理由

私個人の意見として、以下の内容が理由となります

1. 年単位で放置されているIssueが、解決済みであるかどうかの判別がつきにくくなっている
2. 「Issueが多い≒バグが多い」という、印象を持たれる懸念
3. 新しくIssueを出しても、解決されないのではないかという疑念を抱かせてしまう懸念
4. Issueを登録するにしても重複する可能性がある（GitHubのIssue検索が弱いところや、Duplicateを出す方の負担懸念）

### 1について

長期間放置されているIssueは、既に現在のバージョンで解決済みであるかもしれないと考えられます。
そのため、一度クローズしてしまい必要に応じてリオープンしてもよいのではと思いました。

また、仕様が変わっていくMisskeyだと、年単位で古いissueだと現在の仕様と合わない部分が出てきてしまい、Issueが腐ってしまっている可能性があると思いました。

### 2について

Featureラベルもあるため一概に言えないことや、開発者の視点からするとそうではないという認識になると思うのですが、
何も知らない人からするとIssueが多いということはその分不具合が多いのではないか。という不要な心配を与えるのではないかと思いました。
これにより、本当に必要なIssue以外はクローズしてしまったほうが良いのではと考えました。

### 3について

MisskeyのIssueは日を追うごとに少しずつ増えていっている状態です。
もちろん必要があって増えているというのはあると思うのですが、
レスポンスがないまま放置されていってしまっているIssueが多いと考えています。
Issueを登録する側からすると、登録をしても相手にされないまま埋もれていくのではないか。という懸念を与えてしまうと思います。

### 4について

普段からDuplicateからのクローズが発生していますが、報告したい内容で既出のIssueが見つけづらい状態であると考えています。
一部メンバーによってDuplicateのコメントとクローズ対応はされているものの、Duplicateの指摘を入れるメンバーのコストを考えると、整理していったほうが良いと考えました。

## 個人的見解

- コメントが1件もついていない
- 追加情報を求めたが、Issue主から返事がない
- スレッドが長期間止まっている

といったIssueは、一度クローズしてしまっても良いと考えました。
もちろん、スレッドは止まっているが必要な情報のため残している。というケースも有るため条件は考える必要があります。

## まとめ

この件についてsyuiloさんとやりとりをしていたのですが、
「定期的にIssueを見直しているが、outdatedなIssueはそこまでない印象であり、
現在オープンなIssueは必要なものであるため、整理の必要性は感じない。
ただ、他の方がどのように感じているかはわからない。」

という話があったため、syuiloさんの意見が尊重されるべきであると考えていますが、
一度明確にするために他の方々の意見も聞きたいと考え今回このDiscussionを建てました。

自分の挙げた理由がこのプロジェクトに則していない可能性ももちろんあるため、「これは違うよ」ってところがありましたらご指摘いただけると幸いです。

ぜひ皆様の考えを教えていただけないでしょうか。

よろしくお願いいたします。

**Comment by kakkokari-gtyih** [comment]
バージョンの更新によって機能そのものが削除 or 根本的に変更されて、その機能に連れて触れているissueがいわゆる「該当なし」になってることがあるのは何個か見たことがあるのでその辺は閉じても良いかも（ただしその判別が鬼）

**Reply by kakkokari-gtyih** [reply]
実際古いissueに対応するようなPRが今でも出てたりするので一概に閉じればいいかと言われればそれも違う気はする

もしくはPRが付いてない古いやつは一回全部閉じちゃって、「要りそうなもんはもう一回出してください」方式にするか

**Comment by KisaragiEffective** [comment]
古いissueから見ていたのですが、だいたいのissueは以下の4つに分類されるように思います。
1. ふわっとした記載内容→PRがなければ一括stale扱いで良さそう。どのみち何を求めてるかの掘り下げが必要だが、数年前のことなんて誰も覚えているわけがない
2. 概念ごと消滅している→closeできる
3. 現時点では修正されている→closeできる
4. 現時点でも生き残っている→残置されるべき

1の一例として、本文が空、またはbountyを貼ろう！の一文だけというパターンがあるようです。
そこで、試験的にこれらをクローズしたら多少の棚卸しになるのでは？

**Reply by KisaragiEffective** [reply]
[oldest](https://github.com/misskey-dev/misskey/issues?q=is%3Aissue+is%3Aopen+sort%3Acreated-asc) で見ていましたが、outdatedなものを効率良く見るためには[least recently updated](https://github.com/misskey-dev/misskey/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-asc) のほうがいいかも？

**Comment by KisaragiEffective** [comment]
4については検索で引っかかるであろう単語を本文中に入れさせることで将来的には解決できそうです。例えばTypeScriptはこの解決方法を採用しています (<https://github.com/microsoft/TypeScript/blob/d8086f14b6b97c0df34a0cc2f56d4b5926a0c299/.github/ISSUE_TEMPLATE/bug_report.yml>)。

**Reply by zyoshoka** [reply]
duplicate が生じる要因の一つに、英語で書かれた issue が日本語で書かれた issue の重複だった（逆も然り）、といった言語面の問題もあるかと思います。そもそも英語圏の人が日本語だけで書かれた issue から自分の issue が duplicate かどうかを確かめるのは多くの場合厳しいと思うので、こういった search terms 欄を設けた上で、日本語で issue を書く側は（可能な限り）簡易的な英語でのキーワードを併記しておくようにすれば、こうした duplicate も減るのではないでしょうか。

**Comment by tai-cha** [comment]
> コメントが1件もついていない
> 追加情報を求めたが、Issue主から返事がない
> スレッドが長期間止まっている

closeしない方がいいものにopenなPullReqからリンクされている場合があるかもしれないです（本来は十分にIssueで議論されてから実装されるべきですがこれはIssueに対して十分にレスポンスができていない現状を考えると閉じないほうがよさそう）
mergedな場合はおそらく部分的に解決済みかもしれないので追加情報が欲しくなりnot plannedなどでcloseされたPRは関係なく閉じていいはず

**Reply by tai-cha** [reply]
ただPullReq自体も古いとなるとそれ自体がoutdatedな可能性もあるな

**Comment by t1nyb0x** [comment]
皆様ご意見ありがとうございます。
既にDuplicateであったり、概念上消滅または、修正済みであろうIssueがクローズされているのを確認しています。ありがとうございます。

今のところ出ている案で、Kisaragiさんの案をベースに考えると

1. 本文が空、またはbountyを貼ろう！の一文のIssue
 　→PRがなければ一括stale扱い
2. 概念ごと消滅している
 　→close
3. 現時点では修正されている
    →closeできる
4. 現時点でも生き残っている
　→残置されるべき

4については、zyoshokaさんが話されていた今後search terms欄 [参考](https://github.com/microsoft/TypeScript/blob/d8086f14b6b97c0df34a0cc2f56d4b5926a0c299/.github/ISSUE_TEMPLATE/bug_report.yml) を今後テンプレートに取り込むことで、重複Issueを減らせると思いました。


自分が挙げていた

> コメントが1件もついていない
> 追加情報を求めたが、Issue主から返事がない
> スレッドが長期間止まっている

という点はPullReqが既に出ているものだとクローズしにくいが、PullReq自体がoutdatedのパターンもあり得るとtai-chaさんのご意見の通りだと私も思いました。
ここについては、Kisaragiさんの案をベースにしちゃっても良いかと思いました。

クローズしたIssueについて、kakkokariさんの言う通りもし必要であればまた出してくださいまたは、再度開ける形式でも良いのかなと自分も思います。


内容整理してみましたが、認識違うよ！ってのありましたらご指摘いただけると幸いです。

よろしくお願いします。

**Comment by KisaragiEffective** [comment]
ラベルがついていないissueの中には、質問の形式をとっているものがあります。これらのissueはconvert to discussionするべきでしょうか？

**Reply by t1nyb0x** [reply]
BugやFeatureにあたらない、単純な質問であればDiscussionへ移動しちゃっても自分は良いと思います

**Reply by Sayamame-beans** [reply]
issueとDiscussionの使い分けがそのようになっているのであれば、移動して良さそうに思います(Misskeyのその辺りの方針を見直せてないので未確認です)
内容がバグ等に絡みそうなら残しておくのが良さそうですね

**Reply by KisaragiEffective** [reply]
> issueとDiscussionの使い分けがそのようになっている

一般にIssueとDiscussionを併置しているということは、Q&AなどはDiscussionでやってほしいという意図の現れだと考えています。(それがmisskey-dev/misskeyに当てはまるかどうかはさておき…)

**Reply by t1nyb0x** [reply]
Discussionを新しく作る際、カテゴリにQ&Aがあるので元々その意図はあったんじゃないかなぁ・・・と思いました
![image](https://github.com/misskey-dev/misskey/assets/8513531/b8409f74-c693-4939-8994-7654f67cba4b)

**Reply by Sayamame-beans** [reply]
> - Do not use Issues to ask questions or troubleshooting.
>   - Issues should only be used to feature requests, suggestions, and bug tracking.
>   - Please ask questions or troubleshooting in [GitHub Discussions](https://github.com/misskey-dev/misskey/discussions) or [Discord](https://discord.gg/Wp8gVStHW3).

https://github.com/misskey-dev/misskey/blob/c51347d78bc6dd30b6b4db2af64f0ea4bc83091e/CONTRIBUTING.md#issues

良さそうです。

**Reply by KisaragiEffective** [reply]
ラベルなしのものはDiscussionsに移動したり適切なラベルをつけたりして無くなりました 🎉

**Comment by Sayamame-beans** [comment]
issueを機能等に応じてある程度分類したいのですが、Projects(既存になければ作成する)での対処で問題ないでしょうか?
(ラベルはCI系で使っていそうなので増やすと良くないのかなと思っています)

**Reply by t1nyb0x** [reply]
分類分けはProjectsで良いかなと考えてます。
すでに一部は私の方でつけているので、そのやり方のが他の方にとってもわかりやすいのかなと思います。

**Reply by t1nyb0x** [reply]
ラベルは既存のものあれば活用するで良いのかな。たぶん…
ダメな理由がない限りは活用したほうがよさそう。

**Reply by anatawa12** [reply]
CI系で使ってるのはpr周りなので、issueではCIにほぼ影響ないと思いますし既存のは使っていいと思います
